src/crc16.c

```diff
@@ -29,60 +29,77 @@
  * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#define REVERSE_8_BIT(n)                                       \
+    do {                                                       \
+        n = ((n & 0xf0f0f0f0) >> 4) | ((n & 0x0f0f0f0f) << 4); \
+        n = ((n & 0xcccccccc) >> 2) | ((n & 0x33333333) << 2); \
+        n = ((n & 0xaaaaaaaa) >> 1) | ((n & 0x55555555) << 1); \
+    } while (0)
+
+uint8_t reverse_8_bit(uint8_t n) {
+    REVERSE_8_BIT(n);
+    return n;
+}
+
+uint16_t reverse_16_bit(uint16_t n) {
+    n = ((n & 0xff00ff00) >> 8) | ((n & 0x00ff00ff) << 8);
+    REVERSE_8_BIT(n);
+    return n;
+}
+
+/* The lookup table just needs to store every 16th entry
+ * of the standard lookup table.
+ */
+static const uint16_t crc16_half_byte_tbl[16] = {
+    0x0000,
+    0x1081,
+    0x2102,
+    0x3183,
+    0x4204,
+    0x5285,
+    0x6306,
+    0x7387,
+    0x8408,
+    0x9489,
+    0xa50a,
+    0xb58b,
+    0xc60c,
+    0xd68d,
+    0xe70e,
+    0xf78f,
+};
+
+/* Adapted from: https://create.stephan-brumme.com/crc32/
+ * Apply half-byte lookup table algorithm for the balance
+ * between performance and the size of lookup table.
+ */
+uint16_t crc16_base(uint16_t crc, uint8_t v) {
+    crc ^= v;
+    crc = (crc >> 4) ^ crc16_half_byte_tbl[crc & 0x000f];
+    crc = (crc >> 4) ^ crc16_half_byte_tbl[crc & 0x000f];
+    return crc;
+}
+
 /* CRC16 implementation according to CCITT standards.
  *
- * Note by @antirez: this is actually the XMODEM CRC 16 algorithm, using the
- * following parameters:
+ * Note by @cudachen: this is actually the XMODEM CRC-16 algorithm (with
+ * bit-reflected), using the following parameters:
  *
  * Name                       : "XMODEM", also known as "ZMODEM", "CRC-16/ACORN"
  * Width                      : 16 bit
- * Poly                       : 1021 (That is actually x^16 + x^12 + x^5 + 1)
- * Initialization             : 0000
- * Reflect Input byte         : False
- * Reflect Output CRC         : False
- * Xor constant to output CRC : 0000
- * Output for "123456789"     : 31C3
+ * Poly                       : 0x8408 (That is actually 0x1021, or x^16 + x^12 + x^5 + 1)
+ * Initialization             : 0x0000
+ * Reflect Input byte         : True
+ * Reflect Output CRC         : True
+ * Xor constant to output CRC : 0x0000
+ * Output for "123456789"     : 0x31C3
  */
```

Comment:
Why change the description? It the old description not correct or is it equivalent to your new description, but yours is written in a reflected way...?

I think we can keep the old description, if it's not incorrect. It seems to be the most common way to describe this CDC-16 variant. For example, according to this page, the Xmodem variant has the parameters mentioned by Antirez:

* https://reveng.sourceforge.io/crc-catalogue/all.htm#crc.cat.crc-16-xmodem

If a reflected and non-reflected variant are equivalent with just different numbers in the table, is it possible to avoid the `reverse_8_bit()` and `reverse_16_bit()` calls, if we just use different numbers in the table?